MOM standard proposal (#2848)
ethereum/EIPs: EIPS/eip-2848.md

My Own Messages (MOM) is a standard to notarize and share messages using Ethereum, so that contents are verificable and decentralized.

## Abstract
<!--A short (~200 word) description of the technical issue being addressed.-->
My Own Messages (MOM) is a standard to notarize and share messages using Ethereum as a certification layer. This can be achieved just submitting a self-send transaction to the blockchain with a specific format and message data attached.

## Motivation
<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
As a _developer_ or _pool's owner_, I'd like to send messages to my users in a decentralized way. They must be able to easily verify my role in the smart contract context (owner, user, and so on) and they must be able to do it without relying on external, insecure and hackable social media sites (Facebook, Twitter, you name it). Also, I'd like to read messages from my userbase, in the same secure and verifiable manner.

> As a _user_, I want a method to easily share my thoughts and idea, publish content, send messages, receive feedbaks, receive tips, and so on, wihout dealing with any complexity: just write a message, send it and it's done. Also, I want to write to some smart contract's ower or to the sender of some transaction.

Reviewer: ```suggestion
As a _user_, I want a method to easily share my thoughts and idea, publish content, send messages, receive feedbaks, receive tips, and so on, without dealing with any complexity: just write a message, send it and it's done. Also, I want to write to some smart contract's owner or to the sender of some transaction.
```

> As an _explorer service_, I want to give my users an effective way to read informations by smart contract owners and a place to share ideas and informations, without using third party services (i.e. Etherscan uses Disqus, and so on)

Reviewer: ```suggestion
As an _explorer service_, I want to give my users an effective way to read information by smart contract owners and a place to share ideas and information without using third party services (i.e. Etherscan uses Disqus, and so on)
```

It's **RECOMMENDED** that clients inform users about the actual setting of the default content type.

### MOM transactions
Clients **MUST** assume that **invalid MOM transacions don't exist**. If a transaction does not strictly follow the MOM standard, clients **MUST** ignore it and they **MUST NOT** consider it a MOM transaction at all.

Because there can be security implications parsing data sent by users, clients **SHOULD NOT** try to keep track or interpret transactions as _invalid_ MOM transactions.

#### Valid MOM transaction's data structure
| ATTRIBUTE | VALUE |
|:--------|:------------|
| `to` | **MUST** be the same account signing the transaction |
| `value` |  **MUST** be `0` wei |
| `data` | **MUST** be at least `1` byte. The first byte **MUST** be operational code and following bytes **MUST** be based on the operational codes listed below. |

#### List of supported operations and messages
Each operational code has one or more parameters, and all parameters **MUST** be considered mandatory.

> Optional parameters don't exist: if parameters for the specific operational code are not all present or they don't follow the rules, clients **MUST** ignore the transaction completly.

Reviewer: ```suggestion
Optional parameters don't exist: if parameters for the specific operational code are not all present or they don't follow the rules, clients **MUST** ignore the transaction completely.
```

Please also note that, because it's usually up to the author of a message to be sure the content is available to everyone, if a `DELETE` message was sent it's very likely the content referenced by the multihash isn't available anymore, simply because probably it's not shared by anyone.

## Rationale
<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
Ethereum is _account based_, so it's good to be identified as a single source of information.

It is also able of doing notarization very well and to impose some restrictions on transaction's structure, so it's good for commands.

> IPFS, Swarm or other CANs (Content Addressable Networks) or storage methods are good to store lot of informations. So, the union of both worlds it's a good solution to achieve the objectives of this message standard.

Reviewer: ```suggestion
IPFS, Swarm or other CANs (Content Addressable Networks) or storage methods are good to store a lot of information. So, the union of both worlds it's a good solution to achieve the objectives of this message standard.
```

The objective is also to avoid in the first place any kind of scam and malicious behaviors, so MOM don't allow to send transactions to other accounts and the value of a MOM transaction is always 0.

### Why not using a smart contract?
MOM wants to be useful, easy to implement and read, error proof, fast and cheap, but:
- using a smart contract for messages can leads more easily to errors and misunderstandings:
    - address of the contract can be wrong
    - smart contract must be deployed on that specific network to send messages
- executing a smart contract costs much more then sending transactions
- executing a smart contract just to store static data is the best example of an anti-pattern (expensive and almost useless)

Without a specific smart contract to rely on, the MOM standard can be implemented and used right now in any existing networks, and even in future ones.

> Finally, if you can achive exactly the same result without a smart contract, you didn't need a smart contract at the first place.

Reviewer: ```suggestion
Finally, if you can achieve exactly the same result without a smart contract, you didn't need a smart contract at the first place.
```

### Why not storing messages directly on-chain?
There's no benefit to store _static_ messages on-chain, if they are not related to some smart contract's state or if they don't represent exchange of value. The cost of storing data on-chain is also very high.

### Why not storing op codes inside the message?
While cost effectiveness is a very important feature in a blockchain related standard, there's also a compromise to reach with usability and usefulness.

Storing commands inside the messages forces the client to actually download messages to understand what to do with them. This is very unefficent, bandwidth and time consuming.

Being able to see the commands before downloading the content, it allows the client to recreate the history of all messages and then, at the end, download only updated messages.

> Creating a structure for the content of the messages leads to many issues and considerations in parsing the contect, if it's correct, mispelled, and so on.

Reviewer: ```suggestion
Creating a structure for the content of the messages leads to many issues and considerations in parsing the content, if it's correct, misspelled, and so on.
```

```
npm install @internetofpeers/mom-js
```

Transaction [`0x8e49485c56897757a6f2707b92cd5dad06126afed92261b9fe1a19b110bc34e6`](https://etherscan.io/tx/0x8e49485c56897757a6f2707b92cd5dad06126afed92261b9fe1a19b110bc34e6) is an example of a valid MOM transaction already mined on the Mainnet; it's an `ADD` message.

## Security Considerations
<!--All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.-->
MOM is very simple and it has no real security concerns by itself. The standard already considers valid only transactions with `0` value and where `from` and `to` addresses are equals.

> The only concerns can come from the payload, but it is more realted to the client and not to the standard itself, so here you can find some security suggestions related to clients implementing the standard.

Reviewer: ```suggestion
The only concerns can come from the payload, but it is more related to the client and not to the standard itself, so here you can find some security suggestions related to clients implementing the standard.
```
